### ISSUE TITLE
Support for dynamic python loading [manylinux + pip]

### ISSUE THREAD

**Issue by pramodk** [BlueBrain/nmodl#315]
To build portable wheel, we need to take care of following things:

- [x] avoid linking to `libpython.so` for executables (this is added by pybind11)
```
[ 64%] Linking CXX executable ../../bin/nmodl
cd /root/nrn/external/coreneuron/external/nmodl/build/src/nmodl && /opt/cmake/bin/cmake -E cmake_link_script CMakeFiles/nmodl.dir/link.txt --verbose=1
/opt/rh/devtoolset-8/root/usr/bin/c++    -rdynamic CMakeFiles/nmodl.dir/main.cpp.o  -o ../../bin/nmodl -Wl,-rpath,:::::::::::::: ../printer/libprinter.a ../codegen/libcodegen.a ../visitors/libvisitor.a ../symtab/libsymtab.a ../utils/libutil.a ../lexer/liblexer.a -lpython3.7m ../../ext/fmt/libfmt.a
/opt/rh/devtoolset-8/root/usr/libexec/gcc/x86_64-redhat-linux/8/ld: cannot find -lpython3.7m
```
- [x] `libpython.so` should be dynamically loaded by `nmodl` binary 
- [x] `nrnunits.lib` path should be relative to target installation. One can get error like:

```
$ nmodl
terminate called after throwing an instance of 'std::runtime_error'
  what():  Could not found nrnunits.lib
Aborted
```

**Comment by pramodk** [comment]
Copying the discussion from Slack:

```
MH: I'm not understanding why there is a need to explicitly load libpython. \
It was needed for  nrniv because one needed to handle the case where it did \
not exist and when the version was not known at build.

It was needed for python  followed by import neuron because libnrniv had to use \
dlsym to connect itself to python via the libnrnpython...

In you shim it is confusing because the shim presupposes libpython is already loaded. \
So does a second dlopen just mean you need to find some symbols in it?


PK : I thought so. But is that true? because I get following error:

$ nmodl
/home/kumbhar/nrn/v37/lib/python3.7/site-packages/neuron/.data/bin/nmodl: symbol lookup error: /home/kumbhar/nrn/v37/lib/python3.7/site-packages/neuron/.data/bin/nmodl: undefined symbol: _Py_TrueStruct

The actual binary .../.data/bin/nmodl needs to link to libpython.so but it’s not available while \
building wheel. So we use -Wl,--unresolved-symbols=ignore-all flag and that creates nmodl binary:

$ ldd ./v37/lib/python3.7/site-packages/neuron/.data/bin/nmodl
	linux-vdso.so.1 (0x00007ffd129dc000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f5fc044c000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f5fc00ae000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f5fbfc7f000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f5fbfa60000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f5fbf66f000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f5fc07d5000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f5fbf467000)

I also thought it won’t be necessary to libpython.
```

**Comment by nrnhines** [comment]
> @Michael Hines: when starting from python (i.e. mpirun -n 2 python test.py -mpi), doesn’t neuron dlopen python library internally?

No. It definitely loads if python is embedded. ie.
```
      if (!nrn_is_python_extension) {
```
What always is dlopen'd is the proper libnrnpython... . The proper one is known from the precise
value of nrn_is_python_extension which was set in the hoc module (src/nrnpython/inithoc.cpp) derived from ```Py_GetVersion()```

**Comment by pramodk** [comment]
> No. It definitely loads if python is embedded. ie.

It's not clear to me what is meaning of ` if python is embedded. ie. if (!nrn_is_python_extension)` 

Here is what I was trying to understand:

```
$ python
>>> import os
>>> os.getpid()
20250

$ lsof -p 20250
...
      Output information may be incomplete.
COMMAND   PID    USER   FD   TYPE DEVICE SIZE/OFF    NODE NAME
python  20250 kumbhar  cwd    DIR  253,2     4096 4327695 /home/kumbhar/nrn
python  20250 kumbhar  rtd    DIR  253,0     4096       2 /
python  20250 kumbhar  txt    REG  253,0  4873376  701057 /usr/bin/python3.7
python  20250 kumbhar  mem    REG  253,0   170784 1310792 /lib/x86_64-linux-gnu/libtinfo.so.5.9
python  20250 kumbhar  mem    REG  253,0   294632 1311301 /lib/x86_64-linux-gnu/libreadline.so.7.0
python  20250 kumbhar  mem    REG  253,0    31752  701243 /usr/lib/python3.7/lib-dynload/readline.cpython-37m-x86_64-linux-gnu.so
python  20250 kumbhar  mem    REG  253,0  3004464  655382 /usr/lib/locale/locale-archive
python  20250 kumbhar  mem    REG  253,0  1700792 1310748 /lib/x86_64-linux-gnu/libm-2.27.so
python  20250 kumbhar  mem    REG  253,0   116960 1310887 /lib/x86_64-linux-gnu/libz.so.1.2.11
python  20250 kumbhar  mem    REG  253,0   202880 1319091 /lib/x86_64-linux-gnu/libexpat.so.1.6.7
python  20250 kumbhar  mem    REG  253,0    10592 1310763 /lib/x86_64-linux-gnu/libutil-2.27.so
python  20250 kumbhar  mem    REG  253,0    14560 1310747 /lib/x86_64-linux-gnu/libdl-2.27.so
python  20250 kumbhar  mem    REG  253,0   144976 1310759 /lib/x86_64-linux-gnu/libpthread-2.27.so
python  20250 kumbhar  mem    REG  253,0  2030544 1310744 /lib/x86_64-linux-gnu/libc-2.27.so
python  20250 kumbhar  mem    REG  253,0   170960 1310735 /lib/x86_64-linux-gnu/ld-2.27.so
python  20250 kumbhar  mem    REG  253,0    26376  657882 /usr/lib/x86_64-linux-gnu/gconv/gconv-modules.cache
python  20250 kumbhar    0u   CHR  136,3      0t0       6 /dev/pts/3
python  20250 kumbhar    1u   CHR  136,3      0t0       6 /dev/pts/3
python  20250 kumbhar    2u   CHR  136,3      0t0       6 /dev/pts/3
```

Then, loaded library explicitly and that shows libpython:

```
>>> ctypes.cdll.LoadLibrary("/usr/lib/python3.7/config-3.7m-x86_64-linux-gnu/libpython3.7m.so")
<CDLL '/usr/lib/python3.7/config-3.7m-x86_64-linux-gnu/libpython3.7m.so', handle 2438130 at 0x7f4a0c667c90>
.....
$ lsof -p 20250
COMMAND   PID    USER   FD   TYPE DEVICE SIZE/OFF    NODE NAME
python  20250 kumbhar  cwd    DIR  253,2     4096 4327695 /home/kumbhar/nrn
python  20250 kumbhar  rtd    DIR  253,0     4096       2 /
python  20250 kumbhar  txt    REG  253,0  4873376  701057 /usr/bin/python3.7
python  20250 kumbhar  mem    REG  253,0  5075632  658878 /usr/lib/x86_64-linux-gnu/libpython3.7m.so.1.0
python  20250 kumbhar  mem    REG  253,0    31032  660687 /usr/lib/x86_64-linux-gnu/libffi.so.6.0.4
python  20250 kumbhar  mem    REG  253,0   126912  701220 /usr/lib/python3.7/lib-dynload/_ctypes.cpython-37m-x86_64-linux-gnu.so
python  20250 kumbhar  mem    REG  253,0   170784 1310792 /lib/x86_64-linux-gnu/libtinfo.so.5.9
python  20250 kumbhar  mem    REG  253,0   294632 1311301 /lib/x86_64-linux-gnu/libreadline.so.7.0
python  20250 kumbhar  mem    REG  253,0    31752  701243 /usr/lib/python3.7/lib-dynload/readline.cpython-37m-x86_64-linux-gnu.so
python  20250 kumbhar  mem    REG  253,0  3004464  655382 /usr/lib/locale/locale-archive
python  20250 kumbhar  mem    REG  253,0  1700792 1310748 /lib/x86_64-linux-gnu/libm-2.27.so
python  20250 kumbhar  mem    REG  253,0   116960 1310887 /lib/x86_64-linux-gnu/libz.so.1.2.11
python  20250 kumbhar  mem    REG  253,0   202880 1319091 /lib/x86_64-linux-gnu/libexpat.so.1.6.7
python  20250 kumbhar  mem    REG  253,0    10592 1310763 /lib/x86_64-linux-gnu/libutil-2.27.so
python  20250 kumbhar  mem    REG  253,0    14560 1310747 /lib/x86_64-linux-gnu/libdl-2.27.so
python  20250 kumbhar  mem    REG  253,0   144976 1310759 /lib/x86_64-linux-gnu/libpthread-2.27.so
python  20250 kumbhar  mem    REG  253,0  2030544 1310744 /lib/x86_64-linux-gnu/libc-2.27.so
python  20250 kumbhar  mem    REG  253,0   170960 1310735 /lib/x86_64-linux-gnu/ld-2.27.so
python  20250 kumbhar  mem    REG  253,0    26376  657882 /usr/lib/x86_64-linux-gnu/gconv/gconv-modules.cache
python  20250 kumbhar    0u   CHR  136,3      0t0       6 /dev/pts/3
python  20250 kumbhar    1u   CHR  136,3      0t0       6 /dev/pts/3
python  20250 kumbhar    2u   CHR  136,3      0t0       6 /dev/pts/3
```

In shim I was trying to do:

```
    ctypes.cdll.LoadLibrary("/usr/lib/python3.7/config-3.7m-x86_64-linux-gnu/libpython3.7m.so")
    os.execv(exe, sys.argv)
```
But that doesn't work as well. So my guess is that the libraries loaded by python are not inherited when we start subprocess or exec*.

**Comment by nrnhines** [comment]
To me, python embedded means that one launched nrniv and libpython is statically linked or dynamically loaded. Extended means that one launched python and on ```import hoc``` the neuron
shared libraries dynamically loaded.
```
$ lsof -p 20250
...
      Output information may be incomplete.
```
That's interesting that it admits it is incomplete:)  I remember having trouble with lsof when I was
developing nrnpyenv.sh and seeing that it was only occasionally forthcoming, falling back on 'find'
and then falling back further to importing ctypes and looking at the__file__ for every sys.modules.

**Comment by ohm314** [comment]
So, since it's similar in nmodl in that python is embedded in the `nmodl` binary through pybind11. Do we need to add extra code to `dlopen` libpython if we cannot link it at build time?

**Comment by pramodk** [comment]
@ohm314 : I still have some unclarity on this matter : note that for `pybind` we have to do following ([commit](https://github.com/BlueBrain/nmodl/blob/a6b95081e37f30b9802b0d7d9a30bc889d971a99/src/visitors/CMakeLists.txt)): 

```
# ~~~
# pybind11::embed adds PYTHON_LIBRARIES to target_link_libraries. To avoid link to
# libpython, we can use `module` interface library from pybind11.
# ~~~
if(NOT LINK_AGAINST_PYTHON)
  target_link_libraries(visitor PRIVATE module)
else()
  target_link_libraries(visitor PRIVATE pybind11::embed)
endif()
``` 
i.e. we can't embed python because it's not available during build time. I would like to understand / discuss this bit with @ferdonline.

**Comment by ohm314** [comment]
@pramodk - yes I understand and agree with you on this. Makes sense (see also my comments connected with this in that PR). But this build time.. my question is that in runtime, how do we find libpython. You seemed to suggest earlier that by using the shims this would work, but I was not so sure it would - and your comments above seem to also point to us needing to do `dlopen` ? It seems the safer choice anyway, otherwise we'd have to always rely on the shims, even for the non `setup.py` installation.

**Comment by nrnhines** [comment]
Perhaps you would like to avoid but nrniv offloads this issue to nrnpyenv.sh. But that context was
radical in the sense that there was no information about python except that you could launch it
or else a -pyexe arg was required. In the wheel case, it seems that one actually does know all the
necessary info about what python is being used.

**Comment by ohm314** [comment]
Python dynamic library loading has been in implemented in #325
